libsql-sqlite3/src/vector.c

```diff
@@ -0,0 +1,558 @@
+    memset(valueBuf, 0, iBuf);
+    iBuf = 0;
+    if( pVector->type == VECTOR_TYPE_FLOAT32 ){
+      elemsFloat[iElem++] = elem;
+    } else if( pVector->type == VECTOR_TYPE_FLOAT64 ){
+      elemsDouble[iElem++] = elem;
+    }
+    if( this == ']' ){
+      break;
+    }
+  }
+  while( pzText && sqlite3Isspace(*pzText) )
+    pzText++;
+
+  if( pzText != NULL && *pzText != ']' ){
+    *pzErrMsg = sqlite3_mprintf("malformed vector, doesn't end with ']'");
+    goto error;
+  }
+  pzText++;
+
+  while( pzText && sqlite3Isspace(*pzText) )
+    pzText++;
+  
+  if( pzText != NULL && *pzText != '\0' ){
+    *pzErrMsg = sqlite3_mprintf("malformed vector, extra data after closing ']'");
+    goto error;
+  }
+  pVector->dims = iElem;
+  return 0;
+error:
+  return -1;
+}
+
+int vectorParseSqliteBlob(
+  sqlite3_value *arg,
+  Vector *pVector,
+  char **pzErrMsg
+){
+  switch (pVector->type) {
+    case VECTOR_TYPE_FLOAT32: 
+      return vectorF32ParseSqliteBlob(arg, pVector, pzErrMsg);
+    case VECTOR_TYPE_FLOAT64: 
+      return vectorF64ParseSqliteBlob(arg, pVector, pzErrMsg);
+    default: 
+      assert(0);
+  }
+  return -1;
+}
+
+int detectBlobVectorParameters(sqlite3_value *arg, int *pType, int *pDims, char **pzErrMsg) {
+  const u8 *pBlob;
+  int nBlobSize;
+  
+  assert( sqlite3_value_type(arg) == SQLITE_BLOB );
+
+  pBlob = sqlite3_value_blob(arg);
+  nBlobSize = sqlite3_value_bytes(arg);
+  if( nBlobSize % 2 != 0 ){ 
+    // we have trailing byte with explicit type definition
+    *pType = pBlob[nBlobSize - 1];
+  } else { 
+    // else, fallback to FLOAT32
+    *pType = VECTOR_TYPE_FLOAT32;
+  }
+  if( *pType == VECTOR_TYPE_FLOAT32 ){
+    *pDims = nBlobSize / 4;
+  } else if( *pType == VECTOR_TYPE_FLOAT64 ){
+    *pDims = nBlobSize / 8;
+  } else{
+    *pzErrMsg = sqlite3_mprintf("invalid binary vector: unexpected type: %d", *pType);
+    return -1;
+  }
+  if( *pDims > MAX_VECTOR_SZ ){
+    *pzErrMsg = sqlite3_mprintf("invalid binary vector: max size exceeded: %d > %d", *pDims, MAX_VECTOR_SZ);
+    return -1;
+  }
+  return 0;
+}
+
+int detectTextVectorParameters(sqlite3_value *arg, int typeHint, int *pType, int *pDims, char **pzErrMsg) {
+  const u8 *text;
+  int textBytes;
+  int iText;
+  int textHasDigit;
+  
+  assert( sqlite3_value_type(arg) == SQLITE_TEXT );
+  text = sqlite3_value_text(arg);
+  textBytes = sqlite3_value_bytes(arg);
+  if( typeHint == 0 ){ 
+    *pType = VECTOR_TYPE_FLOAT32;
+  }else if( typeHint == VECTOR_TYPE_FLOAT32 ){
+    *pType = VECTOR_TYPE_FLOAT32;
+  }else if( typeHint == VECTOR_TYPE_FLOAT64 ){
+    *pType = VECTOR_TYPE_FLOAT64;
+  }else{
+    *pzErrMsg = sqlite3_mprintf("unexpected vector type");
+    return -1;
+  }
+  *pDims = 0;
+  for(iText = 0; iText < textBytes; iText++){
+    if( text[iText] == ',' ){
+      *pDims += 1;
+    }
+    if( sqlite3Isdigit(text[iText]) ){
+      textHasDigit = 1;
+    }
+  }
+  if( textHasDigit ){
+    *pDims += 1;
+  }
```

Comment:
Why are we adding one more dimension if there's a digit in the text?

Reply:
To distinguish between zero and one dimensional vectors (both have no commas)

Reply:
Ok I can see that now. Thanks!